hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java

```diff
@@ -54,192 +57,208 @@
+
+        // Using the executor service, submit a task for each transaction in the event.
+        final var tasks = new ArrayList<WorkItem>(1000); // Some arbitrary number
+        while (transactions.hasNext()) {
+            // Skip platform (system) transactions.
+            final var platformTx = transactions.next();
+            if (platformTx.isSystem()) continue;
 
-    public void preHandle(@NonNull final Iterator<Transaction> itr, @NonNull final HederaState state) {
-        // Each transaction in the event will go through pre-handle using a background thread
-        // from the executor service. The Future representing that work is stored on the
-        // platform transaction. The HandleTransactionWorkflow will pull this future back
-        // out and use it to block until the pre handle work is done, if needed.
-        final ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
-        while (itr.hasNext()) {
-            final var platformTx = itr.next();
-            final var future = runner.apply(() -> {
-                try {
-                    final var metadata = securePreHandle(state, platformTx);
-                    platformTx.setMetadata(metadata);
-                } catch (final PreCheckException e) {
-                    platformTx.setMetadata(createInvalidResult(e.responseCode()));
-                }
-            });
-            futures.add(future);
+            // Submit the task to the executor service and put the resulting Future as the metadata on the transaction
+            final var future =
+                    exe.submit(() -> preHandleTransaction(creator, readableStoreFactory, accountStore, platformTx));
+            tasks.add(new WorkItem(platformTx, future));
         }
 
-        // wait until all transactions were processed before returning
-        final CompletableFuture<?>[] array = futures.toArray(new CompletableFuture<?>[0]);
-        CompletableFuture.allOf(array).join();
+        // Waits for all the background threads to complete their work and stores the resulting PreHandleResult
+        // as the transaction's metadata.
+        for (final var task : tasks) {
+            try {
+                // FUTURE: Enhance this so the timeout duration is a configuration property (and then update the
+                // unit tests, so they use a much shorter configuration, so test execution is faster).
+                final var result = task.future.get(1, TimeUnit.SECONDS);
+                task.platformTx.setMetadata(result);
+            } catch (InterruptedException e) {
+                // The thread should only ever be interrupted during shutdown, so we can just log the error.
+                logger.error("Interrupted while waiting for a transaction to be pre handled.", e);
+                Thread.currentThread().interrupt();
+                task.platformTx.setMetadata(unknownFailure());
```

Comment:
Should we have a different failure for `InterruptedException` because its not UNKNOWN ? Does this also mean we return UNKNOWN error code in `HandleException` also whenever we receive `InterruptedException`? 
I personally don't like `UNKNOWN` since it doesn't provide any info

Reply:
I think we just need a status code that indicates that "something bad happened, it doesn't really matter what, try again later". I'm using UNKNOWN for that, but maybe there is a better name for it. But I don't think it is useful to disambiguate what exactly happened (InterruptedException vs. some other Throwable) unless we're going to take distinct action for those different types. At the moment I don't think we will (?) but if we do, I agree it should be clarified here.